bpf/aya-bpf/src/maps/xdp/cpu_map.rs

```diff
@@ -0,0 +1,124 @@
+///     MAP.redirect(7, xdp_action::XDP_DROP as u64)
+/// }
+/// ```
+#[repr(transparent)]
+pub struct CpuMap {
+    def: UnsafeCell<bpf_map_def>,
+}
+
+unsafe impl Sync for CpuMap {}
+
+impl CpuMap {
+    /// Creates a [`CpuMap`] with a set maximum number of elements.
+    ///
+    /// In a CPU Map, an entry represents a CPU core. Thus there should be as many entries as there
+    /// are CPU cores on the system. It can be set to zero here, and updated by userspace at
+    /// runtime. Refer to the userspace documentation for more information.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use aya_bpf::{macros::map, maps::CpuMap};
+    ///
+    /// #[map]
+    /// static MAP: CpuMap = CpuMap::with_max_entries(8, 0);
+    /// ```
+    pub const fn with_max_entries(max_entries: u32, flags: u32) -> CpuMap {
+        CpuMap {
+            def: UnsafeCell::new(bpf_map_def {
+                type_: BPF_MAP_TYPE_CPUMAP,
+                key_size: mem::size_of::<u32>() as u32,
+                value_size: mem::size_of::<bpf_cpumap_val>() as u32,
+                max_entries,
+                map_flags: flags,
+                id: 0,
+                pinning: PinningType::None as u32,
+            }),
+        }
+    }
+
+    /// Creates a [`CpuMap`] with a set maximum number of elements that can be pinned to the BPF
+    /// File System (bpffs).
+    ///
+    /// See [`CpuMap::with_max_entries`] for more information.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use aya_bpf::{macros::map, maps::CpuMap};
+    ///
+    /// #[map]
+    /// static MAP: CpuMap = CpuMap::pinned(8, 0);
+    /// ```
+    pub const fn pinned(max_entries: u32, flags: u32) -> CpuMap {
+        CpuMap {
+            def: UnsafeCell::new(bpf_map_def {
+                type_: BPF_MAP_TYPE_CPUMAP,
+                key_size: mem::size_of::<u32>() as u32,
+                value_size: mem::size_of::<bpf_cpumap_val>() as u32,
+                max_entries,
+                map_flags: flags,
+                id: 0,
+                pinning: PinningType::ByName as u32,
+            }),
+        }
+    }
+
+    /// Redirects the current packet on the CPU at `index`.
+    ///
+    /// The lower two bits of `flags` are used for the return code if the map lookup fails, which
+    /// can be used as the XDP program's return code if a CPU cannot be found.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use aya_bpf::{bindings::xdp_action, macros::{map, xdp}, maps::CpuMap, programs::XdpContext};
+    ///
+    /// #[map]
+    /// static MAP: CpuMap = CpuMap::with_max_entries(8, 0);
+    ///
+    /// #[xdp]
+    /// fn xdp(_ctx: XdpContext) -> u32 {
+    ///     // Redirect to CPU 7 or drop packet if no entry found.
+    ///     MAP.redirect(7, xdp_action::XDP_DROP as u64)
+    /// }
+    /// ```
+    #[inline(always)]
+    pub fn redirect(&self, index: u32, flags: u64) -> u32 {
```

Comment:
Shouldn't this return Result?

The kernel docs:
```
 Return XDP_REDIRECT on success, or the value of the two
                     lower bits of the flags argument on error.
```

surely we don't want to expose people to _this_?

Reply:
I don't really know, this easily allows to select which action to perform when not found by directly returning the value from redirect:

```rust
#[xdp]
fn xdp(_ctx: XdpContext) -> i32 {
    MAP.redirect(0, XDP_DROP as u64) as i32
}
```

compared to Result, it's a tad different. The most ergonomic way I end up writing probes in rust is to have my `try_xxx` function not return Result but Option, and returning xdp_pass on None. This way, it's really quick to drill down to packets that interest you (eg. eth + ipv4 + udp + vxlan) and let the rest pass (e.g. bgp), and finish the function with `Some(MAP.redirect(0, XDP_DROP as u64) as i32)`. I save a lot repetition on the default action I want.

Reply:
Yes but that's a C API :P A rusty equivalent would be

```
MAP.redirect(0, 0).or(Ok(XDP_DROP))
```

And if you make it return Result<i32, i32> I guess this also works

```
MAP.redirect(0, XDP_DROP as u64)
```

and then in the caller

```
match try_thing(ctx) {
	Ok(res) => res,
	Err(res) => res,
}
```

So I don't see a reason not to return Result?